csharp/src/Apache.Arrow/C/CArrowSchemaExporter.cs

```diff
@@ -0,0 +1,268 @@
+                    // Same prefix as Time32, but allowed time units are different.
+                    return String.Format("tt{0}", FormatTimeUnit(timeType.Unit));
+                // Timestamp
+                case TimestampType timestampType:
+                    return String.Format("ts{0}:{1}", FormatTimeUnit(timestampType.Unit), timestampType.Timezone);
+                // Nested
+                case ListType _: return "+l";
+                case StructType _: return "+s";
+                // Dictionary
+                case DictionaryType dictionaryType:
+                    return GetFormat(dictionaryType.IndexType);
+                default: throw new NotImplementedException($"Exporting {datatype.Name} not implemented");
+            };
+        }
+
+        private static long GetFlags(IArrowType datatype, bool nullable = true)
+        {
+            long flags = 0;
+
+            if (nullable)
+            {
+                flags |= CArrowSchema.ArrowFlagNullable;
+            }
+
+            if (datatype is DictionaryType dictionaryType)
+            {
+                if (dictionaryType.Ordered)
+                {
+                    flags |= CArrowSchema.ArrowFlagDictionaryOrdered;
+                }
+            }
+
+            // TODO: when we implement MapType, make sure to set the KEYS_SORTED flag.
+            return flags;
+        }
+
+        private static unsafe CArrowSchema** ConstructChildren(IArrowType datatype, out long numChildren)
+        {
+            if (datatype is NestedType nestedType)
+            {
+                IReadOnlyList<Field> fields = nestedType.Fields;
+                int numFields = fields.Count;
+                numChildren = numFields;
+
+                unsafe
+                {
+                    var pointerList = (CArrowSchema**)Marshal.AllocHGlobal(numFields * IntPtr.Size);
+
+                    for (var i = 0; i < numChildren; i++)
+                    {
+                        CArrowSchema* cSchema = CArrowSchema.New();
+                        ExportField(fields[i], cSchema);
+                        pointerList[i] = cSchema;
+                    }
+
+                    return pointerList;
+                }
+
+            }
+            else
+            {
+                numChildren = 0;
+                return null;
+            }
+        }
+
+        private static unsafe CArrowSchema* ConstructDictionary(IArrowType datatype)
+        {
+            if (datatype is DictionaryType dictType)
+            {
+                CArrowSchema* cSchema = CArrowSchema.New();
+                ExportType(dictType.ValueType, cSchema);
+                return cSchema;
+            }
+            else
+            {
+                return null;
+            }
+        }
+
+        private static unsafe void ReleaseCArrowSchema(CArrowSchema* schema)
+        {
+            if (schema == null) return;
+            if (schema->release == null) return;
```

Comment:
Why is this check necessary?

Reply:
`release` is set to null in a release function to indicate it has already been released. This check makes it idempotent.

See: https://arrow.apache.org/docs/format/CDataInterface.html#release-callback-semantics-for-producers

Reply:
Well, if the release callback is already null, the consumer cannot have called this function...